src/modules/ekf2/EKF/wind.cpp

```diff
@@ -0,0 +1,104 @@
+ *    without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ * COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
+ * OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ ****************************************************************************/
+
+/**
+ * @file wind.cpp
+ * Helper functions for wind states
+ */
+
+#include "ekf.h"
+#include <ekf_derivation/generated/compute_wind_init_and_cov_from_wind_speed_and_direction.h>
+
+void Ekf::resetWindToExternalObservation(float wind_speed, float wind_direction, float wind_speed_accuracy, float wind_direction_accuracy)
+{
+	const float wind_speed_constrained = math::max(wind_speed, 0.0f);
+
+	// wind direction is given as azimuth where wind blows FROM, we need direction where wind blows TO
+	const float wind_direction_rad = wrap_pi(math::radians(wind_direction) + M_PI_F);
+
+	const float wind_direction_var = sq(math::radians(wind_direction_accuracy));
+	const float wind_speed_var = sq(wind_speed_accuracy);
+
+	matrix::SquareMatrix<float, 2> P_wind;
+	Vector2f wind;
+
+	sym::ComputeWindInitAndCovFromWindSpeedAndDirection(wind_speed_constrained, wind_direction_rad, wind_speed_var, wind_direction_var, &wind, &P_wind);
+
+	const Vector2f wind_var = P_wind.diag();
+
+	ECL_INFO("reset wind states to external observation");
+	_information_events.flags.reset_wind_to_ext_obs = true;
+
+	resetWindTo(wind, wind_var);
+}
+
+void Ekf::resetWindTo(const Vector2f &wind, const Vector2f &wind_var)
+{
+	_state.wind_vel = wind;
+
+	if (PX4_ISFINITE(wind_var(0))) {
+		P.uncorrelateCovarianceSetVariance<1>(State::wind_vel.idx, math::max(sq(_params.initial_wind_uncertainty), wind_var(0)));
+	}
+
+	if (PX4_ISFINITE(wind_var(1))) {
+		P.uncorrelateCovarianceSetVariance<1>(State::wind_vel.idx + 1, math::max(sq(_params.initial_wind_uncertainty), wind_var(1)));
+	}
+}
+
+void Ekf::resetWindCov()
+{
+	// start with a small initial uncertainty to improve the initial estimate
+	P.uncorrelateCovarianceSetVariance<State::wind_vel.dof>(State::wind_vel.idx, sq(_params.initial_wind_uncertainty));
+}
+
+void Ekf::resetWind()
+{
+#if defined(CONFIG_EKF2_AIRSPEED)
+	if (_control_status.flags.fuse_aspd && isRecent(_airspeed_sample_delayed.time_us, 1e6)) {
+		resetWindUsingAirspeed(_airspeed_sample_delayed);
```

Comment:
It would be good to fully untangle this one and not even need to store `_airspeed_sample_delayed)`. A wind reset using airspeed can then happen from the airspeed control logic when that new sample is popped.

Reply:
Then how would you handle it when the reset is requested from sideslip_fusion or drag_fusion. We could check in there if airspeed is enabled. If so, we set the corresponding flag, if not we reset to zero. But I dont know if there's a benefit in doing so compared to the current implementation.

Reply:
I would argue that sideslip or drag fusion shouldn't be allowed to call a reset to airspeed. They could either call a reset to 0 or just a covariance reset/bump